src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java

```diff
@@ -1,31 +1,255 @@
+        Map<String, String> additions = new HashMap<>();
+        Map<String, String> deletions = new HashMap<>();
+        Map<String, String> updates = new HashMap<>();
+
+        if (overrides.isEmpty() && overrideWar == null) {
+            effectiveArtifacts = getProjectArtfacts();
+        } else {
+            /*
+             * WARNING: Under no circumstances should this code ever be executed when performing a
+             * release.
+             */
+
+            // Create a shadow project for dependency analysis.
+            MavenProject shadow = project.clone();
+
+            // Stash the original resolution for use later.
+            Map<String, String> originalResolution = new HashMap<>();
+            for (Artifact artifact : shadow.getArtifacts()) {
+                originalResolution.put(toKey(artifact), artifact.getVersion());
+            }
+
+            // First pass: apply the overrides specified by the user.
+            applyOverrides(overrides, bundledPlugins, shadow, getLog());
+
+            if (useUpperBounds) {
+                /*
+                 * Do upper bounds analysis. Upper bounds analysis consumes the model directly and
+                 * not the resolution of that model, so it is fine to invoke it at this point with
+                 * the model having been updated and the resolution having been cleared.
+                 */
+                DependencyNode node;
+                try {
+                    ProjectBuildingRequest buildingRequest = new DefaultProjectBuildingRequest(session.getProjectBuildingRequest());
+                    buildingRequest.setProject(shadow);
+                    ArtifactFilter filter = null; // Evaluate all scopes
+                    node = dependencyCollectorBuilder.collectDependencyGraph(buildingRequest, filter);
+                } catch (DependencyCollectorBuilderException e) {
+                    throw new MojoExecutionException("Failed to analyze dependency tree for useUpperBounds", e);
+                }
+                RequireUpperBoundDepsVisitor visitor = new RequireUpperBoundDepsVisitor();
+                node.accept(visitor);
+                Map<String, String> upperBounds = visitor.upperBounds();
+
+                // Second pass: apply the results of the upper bounds analysis.
+                applyOverrides(upperBounds, Collections.emptyMap(), shadow, getLog());
+            }
+
+            /*
+             * At this point, the model has been updated as the user has requested. We now redo
+             * resolution and compare the new resolution to the original in order to account for
+             * updates to transitive dependencies that are not present in the model. Anything that
+             * was updated in the new resolution needs to be updated in the test classpath. Anything
+             * that was removed in the new resolution needs to be removed from the test classpath.
```

Comment:
Yes, exactly. You can see this in action by finding a plugin with an older core baseline (e.g., `text-finder`) and testing with this PR and a megawar of a very new core. The deletions list is a nice summary of all the old JARs I have been removing from core over the past year.

Reply:
Oh this applies to the `jenkins-core` dep too. Interesting, though if you are required to pass `-Djenkins.version=…` anyway, would that not already be making that change to the original project model?

Reply:
> Oh this applies to the `jenkins-core` dep too. Interesting, though if you are required to pass `-Djenkins.version=…` anyway, would that not already be making that change to the original project model?

It would... now that I think about it, I went through that particular debugging exercise before I had added the requirement for `-Djenkins.version`, so I guess you wouldn't be able to replicate it now. :smile: